anitya/lib/librariesio_consumer.py

```diff
@@ -0,0 +1,205 @@
+
+    """
+
+    topic = [
+        'org.fedoraproject.prod.sse2fedmsg.librariesio',
+    ]
+
+    config_key = 'anitya.libraryio.enabled'
+
+    def __init__(self, hub):
+        # If we're in development mode, rewrite some of our topics so that
+        # local playback with fedmsg-dg-replay works as expected.
+        if hub.config['environment'] == 'dev':
+            # Keep the original set, but append a duplicate set for local work
+            prefix, env = hub.config['topic_prefix'], hub.config['environment']
+            self.topic = self.topic + [
+                '.'.join([prefix, env] + topic.split('.')[3:])
+                for topic in self.topic
+            ]
+
+        super(LibrariesioConsumer, self).__init__(hub)
+
+        if not self._initialized:
+            return
+
+        # Initialize fedmsg in this thread's context.
+        hostname = socket.gethostname().split('.', 1)[0]
+        fedmsg.init(name='anitya-libraryio.%s' % hostname)
+        fedmsg.meta.make_processors(**self.hub.config)
+
+    def consume(self, message):
+        """
+        This method is called when a message with a topic this consumer
+        subscribes to arrives.
+
+        Args:
+            message (dict): The fedmsg to process.
+        """
+        project = self._fedmsg_to_project(message)
+        if project is not None:
+            # Double check the validity of the update
+            if self._validate_update(project, message):
+                session = Session()
+                session.add(project)
+                session.commit()
+                Session.remove()
+
+    def _validate_update(self, project, message):
+        """
+        Validate the project was updated.
+
+        This serves two purposes. Firstly, we connect to libraries.io over HTTP
+        so we don't have any authentication. Secondly, we might map the
+        libraries.io project to the wrong Anitya project, so this is a good
+        opportunity to catch and log those problems.
+        """
+        librariesio_version = message['body']['msg']['data']['version']
+        try:
+            anitya_version = get_version(project)
+        except exceptions.AnityaPluginException as e:
+            # TODO what to do here? Bail? Trust libraries.io?
+            _log.warning('libraries.io found an update for %r, but we failed with %s',
+                         project, str(e))
+            return False
+
+        # TODO handle things like 'v' prefix. Should be a general version normalizer
+        # used everywhere.
+        # TODO handle when we find a different version than librariesio. Create the project anyway?
+        equivalent_versions = (librariesio_version == anitya_version)
+        if not equivalent_versions:
+            _log.info('libraries.io found %r had a latest version of %s, but Anitya found %s',
+                      project, librariesio_version, anitya_version)
+        return equivalent_versions
+
+    def _fedmsg_to_project(self, message):
```

Comment:
I'd suggest calling this `_fedmsg_to_candidate_project_update`, as it does more than just work out which project the message relates to:

- for an unknown project, it creates a candidate project instance
- for an existing project, it speculatively updates the latest version info
- in both cases, it *avoids* committing the change, until after the legitimacy of the info can be checked